feature: Documentation command

xNok opened the pull request:
I found myself wanting to generate reference documentation for my policies, and I figured it would be nice if this was a built-in feature of conftest.

While the way of structuring the documentation can be opinionated, I think using the [FlatAnnotationsRefSet](https://www.openpolicyagent.org/docs/latest/policy-language/#go-api) is a safe bet, as OPA's core does all the heavy lifting of organizing annotations and rules.

My initial attempt was to combine `FlatAnnotationsRefSet` and Go-template and offer the least opinionated approach to documentation format that can be easily overridden. However, I had some trouble properly formatting titles (H1, H2, H3, etc...) and had to introduce some custom logic to handle it. (As a side note this could potentially be handles by adding template function libraries such as [sprig](https://github.com/Masterminds/sprig))

boranx commented on document/template.go:
```diff
@@ -0,0 +1,88 @@
+	return nil
+}
+
+func renderTemplate(tpl *TemplateConfig, args interface{}, out io.Writer) error {
```
Could you help understand the motivation behind leaving args as `interface{}` instead of declaring the type concretely (args []Section)

xNok replied:
I don't see any reason to constrain the type here, the underlying `t.Execute` accepts an `interface{}`. The way I see it, this is a wrapper function around `template.Template` that handles either the template is embedded or from the host.

boranx commented on document/metadata.go:
```diff
@@ -0,0 +1,101 @@
+	Annotations *ast.Annotations
+}
+
+func (s Section) Equal(s2 Section) bool {
```
I guess this is not being used, is it?

xNok replied:
It is used in the tests; I will add a comment to explain. In short, `cmp.Equal` will use the `Equal` method of a struct to assess equality instead of deep equality. The `*ast.Annotations` are too complex, and we don't want to have to validate all the fields to infer equality

boranx commented on document/metadata.go:
```diff
@@ -0,0 +1,101 @@
+	return as, nil
+}
+
+type Section struct {
```
some comments explaining `Section` and its attributes might be helpful as it's the backbone of the generation.

boranx commented on document/metadata_test.go:
```diff
@@ -0,0 +1,268 @@
+	}
+
+	as, err := ast.BuildAnnotationSet(parsed)
+	require.Nil(t, err)
```
so far we have followed manual comparisons in testing files, which was also recommended by current maintainers/creators.
Though I'm ok with both, if we end up using it, this needs to be included into `go.mod` as well

boranx commented on internal/commands/document.go:
```diff
@@ -0,0 +1,77 @@
+				}
+				defer func(file *os.File) {
+					if err := file.Close(); err != nil {
+						log.Fatalln(err)
```
I wonder why we don't wanna lean on `return fmt.Errorf` here?

xNok replied:
You can't return from a defer function; at most, you can alter the existing return. But this gets awkward because you need a local variable that you are sure to return.

```go
var err error
//[...]
err = errors.Join(err, filecloeErr)
//[...]
return err
```

It also means that you can't do `return fmt.Errorf("invalid outDir: %s", err)` anymore you have to force yourself to always `return err`. The file not closing is rare enough to log the error and exit.

boranx reviewed (commented):
hey @xNok

Thanks for bringing this together, I actually very liked the idea of using go templates and embedding both so if the user has their own, overriding it to balance flexibility with functionality.
This can help us collectively gather feedback on how users actually use it and refactor to get it closer to the user needs.
+1 from me on implementation.
Just left some comments at first glance, will look at more.

xNok commented on document/metadata.go:
```diff
@@ -0,0 +1,114 @@
+	Annotations *ast.Annotations
+}
+
+// Equal is only relevant for test ans asset that two ection are partially Equal
```
```suggestion
// Equal is only relevant for tests and assert that two sections are partially Equal
```

xNok commented on document/metadata.go:
```diff
@@ -0,0 +1,114 @@
+// Document represent a page of the documentation
+type Document []Section
+
+// Section sequencial piece of documention comprise of ast.Annotations and some pre-processed fields
```
```suggestion
// Section is a sequential piece of documentation comprised of ast.Annotations and some pre-processed fields
```

xNok commented on document/metadata.go:
```diff
@@ -0,0 +1,114 @@
+	return false
+}
+
+// ConvertAnnotationsToSections generate a more convenient struct that can be used to generate the doc
+// First concern is to build a coherent title structure, the ideal case is that each package and each rule as a doc,
+// but this is not guaranteed. I couldn't find a way to call strings.Repeat inside go-template, thus the title symbol is
+// directly provided as markdown (#, ##, ###, etc.)
+// Second the attribute Path of ast.Annotations are not easy to used on go-template, thus we extract it as a string
```
```suggestion
// ConvertAnnotationsToSections generates a more convenient struct that can be used to generate the doc
// First concern is to build a coherent title structure; the ideal case is that each package and each rule has a doc,
// but this is not guaranteed. I couldn't find a way to call `strings.Repeat` inside go-template; thus, the title symbol is
// directly provided as markdown (#, ##, ###, etc.)
// Second, the attribute Path of ast.Annotations are not easy to use on go-template; thus, we extract it as a string
```

xNok commented on document/template.go:
```diff
@@ -0,0 +1,93 @@
+
+type RenderDocumentOption func(*TemplateConfig)
+
+// WithTemplate is a functional option to override the documentation template
+// when overriding the template we assume it is located on the host file system
```
```suggestion
// WithTemplate is a functional option to override the documentation template
// When overriding the template, we assume it is located on the host file system
```

jalseth reviewed (changes requested):
Thanks for the PR! It is very thorough and including documentation on how to use the feature is much appreciated.

xNok commented on document/metadata.go:
```diff
@@ -0,0 +1,114 @@
+	// and smoothed such that subsequent section only defer by +/- 1
+	Depth int
+	// H represent the markdown title symbol #, ##, ###, etc. (produced by strings.Repeat("#", depth))
+	H string
```
doing the operation in go template is pretty ugly, i didn't find a nice way as you can only loop on slices by default

xNok commented:
@jalseth, I have addressed your comment. It seems there is a broken test, and it fails on `master` and my last heavy_check_mark: revision. I have no clue what could be causing it to fail.

![image](https://github.com/user-attachments/assets/83b8e23c-16a5-4e39-b8f5-09f0ca3562f4)

![image](https://github.com/user-attachments/assets/e4fd05c9-99a7-47a0-94e8-ac84cea37c1a)

boranx commented:
@xNok Please rebase the branch, the CI should now be passing

boranx reviewed (approved):
LGTM, thanks for crafting this feature 🎉 
We may further shape it based on user feedback as it's used more and more, but for now it's good to go.

kindly pinging @jalseth / @jpreese for sign-off

mrueg commented:
fwiw, this feature looks similar to https://github.com/plexsystems/konstraint/blob/main/docs/cli/konstraint_doc.md maybe it makes sense to integrate the feature from konstraint?

xNok commented:
> fwiw, this feature looks similar to https://github.com/plexsystems/konstraint/blob/main/docs/cli/konstraint_doc.md maybe it makes sense to integrate the feature from konstraint?

Kubetmetes are not the only thing you can validate with conftest, we use it for terraform and Backstage manifest, in which cases it makes no sense to use Konstraint

anderseknert commented:
This is neat! Thanks for working on this. Let's have it merged!

boranx commented:
Finally getting this over the finish line—appreciate everyone's input!
